### ISSUE TITLE
LocalVariableName should allow second character to be an upper case ASCII letter

### ISSUE THREAD

**Issue by andersthorbeck** [checkstyle/checkstyle#5504]
The checkstyle predefined config file `google_checks.xml` disallows the second character of a variable name to be upper case.

Presumably, this is to enforce the [Google styleguide rule](https://google.github.io/styleguide/javaguide.html#s5.1-identifier-names) of "no special prefixes"
> In Google Style special prefixes or suffixes, like those seen in the examples `name_`, `mName`, `s_name` and `kName`, are not used.

This rule is to avoid prefixes which indicate what the type (int, bool, string, etc) or role (input variable, output variable, etc) or scope (field, local, etc) of the variable is.

However, there are also valid reasons to have a variable name consisting of a single lowercase letter followed immediately by an upper case letter. An example from my personal experience is `eId`, which is an abbreviation of "expression ID". Here, the `e` prefix is not a _"special"_ prefix, like the ones disallowed by the google style sheet, but rather a part of the abbreviation describing the variable.

I believe the `LocalVariableName` checkstyle module should have its format redefined, to allow such cases. A regular expression which would suffice for this would be `"^[a-z][a-zA-Z0-9]*$"`.

--------------

Current behaviour:

`andtho:/tmp$ javac Test.java`:

```
```

`andtho:/tmp$ cat Test.java`:

```
public class Test {
  /** Main method. */
  public static void main(String[] args) {
    // eId abbreviates "expression ID"
    String eId = "Hello world!";
    System.out.println(eId);
  }
}
```

`andtho:/tmp$ curl https://raw.githubusercontent.com/checkstyle/checkstyle/master/src/main/resources/google_checks.xml > config.xml`
`andtho:/tmp$ cat config.xml`:

```
…
<module name = "Checker">
    …
    <module name="TreeWalker">
        …
        <module name="LocalVariableName">
            <property name="tokens" value="VARIABLE_DEF"/>
            <property name="format" value="^[a-z]([a-z0-9][a-zA-Z0-9]*)?$"/>
            <message key="name.invalidPattern"
             value="Local variable name ''{0}'' must match pattern ''{1}''."/>
        </module>
        …
    </module>
</module>
```

`andtho:/tmp$ java -Duser.language=en -Duser.country=US -jar checkstyle-8.8-all.jar -c config.xml Test.java`:

```
Starting audit...
[WARN] /private/tmp/Test.java:5:12: Local variable name 'eId' must match pattern '^[a-z]([a-z0-9][a-zA-Z0-9]*)?$'. [LocalVariableName]
Audit done.
```

---------------

Expected behaviour:

`andtho:/tmp$ java -Duser.language=en -Duser.country=US -jar checkstyle-8.8-all.jar -c config.xml Test.java`:

```
Starting audit...
Audit done.
```

**Comment by rnveach** [comment]
> mName
> eId
> the e prefix is not a "special" prefix, like the ones disallowed by the google style sheet, but rather a part of the abbreviation describing the variable.

The guide doesn't define what is a special prefix and what isn't, so how can you be sure yours isn't considered special by them?
To me, both look the same.

We also only follow guide, so to make this change google style would have to clarify what a special prefix is and update their guide.

**Comment by andersthorbeck** [comment]
In the Google styleguide, they use the term "*special* prefix", they don't just say "prefix". This rule clearly does not apply to _all_ prefixes. It seems obvious to me that the _special_ prefixes they disallow refer to such prefixes as promoted by [Hungarian notation](https://en.wikipedia.org/wiki/Hungarian_notation#Examples), though the styleguide do not explicitly say so.

Note that all their examples of variable names with _special_ prefixes (or suffixes) contain `name` as a substring of the variable name. It seems clear to me that `name` in this case refers to the semantic variable name, _excluding_ information such as variable type etc. such as might be defined with a prefix in Hungarian notation.

My example of `eId` was meant to illustrate a variable name whose prefix is _not_ special, and should thus be allowed. Though, given that this variable name is specific to the business domain in which I currently work, perhaps that was a poor example.

Consider instead a variable containing an [XPath expression](https://www.w3schools.com/xml/xpath_intro.asp) (where two consecutive upper case characters is the official capitalization), which we might reasonably name `xPath`. This variable name, which ought to be perfectly valid, is also discarded as invalid by current checkstyle rules conforming to the google java styleguide.

---

`andtho:/tmp$ cat Test.java`:

```java
public class Test {
  /** Main method. */
  public static void main(String[] args) {
    String xPath = "Hello world!";
    System.out.println(xPath);
  }
}
```

`andtho:/tmp$ java -Duser.language=en -Duser.country=US -jar checkstyle-8.8-all.jar -c config.xml Test.java`:

```
Starting audit...
[WARN] /private/tmp/cs/Test.java:4:12: Local variable name 'xPath' must match pattern '^[a-z]([a-z0-9][a-zA-Z0-9]*)?$'. [LocalVariableName]
Audit done.
```

**Comment by rnveach** [comment]
I have opened an issue with the style team at https://github.com/google/styleguide/issues/326 .
Like I stated we try to follow the guide as closely as possible, so they need to explain what is a special prefix/suffix before we can make a judgement.

> XPath
> xPath

https://google.github.io/styleguide/javaguide.html#s5.3-camel-case
Wouldn't this more follow their example of `IPv6`? In their example, with it following behind another word, it was written in the form `Ipv6`, which I assume would be written as `ipv6` if it was alone.

**Comment by romani** [comment]
> Presumably, this is to enforce the Google styleguide rule of "no special prefixes"

yes, it was do for this. Unlikely it will work for all cases in real life. So some suppression and changes regexp might be required to skip well-known/good-enough in team/project names.
